runtimes/frameless-runtime/src/lib.rs

```diff
@@ -0,0 +1,312 @@
+		fn apply_extrinsic(extrinsic: <Block as BlockT>::Extrinsic) -> ApplyExtrinsicResult {
+			if_std!{
+				println!("Entering apply_extrinsic");
+			}
+
+			let previous_state = sp_io::storage::get(&BOOLEAN_KEY)
+				.map(|bytes| <bool as Decode>::decode(&mut &*bytes).unwrap_or(false))
+				.unwrap_or(false);
+
+			let next_state = match (previous_state, extrinsic) {
+				(_, FramelessTransaction::Set) => true,
+				(_, FramelessTransaction::Clear) => false,
+				(prev_state, FramelessTransaction::Toggle) => !prev_state,
+			};
+
+			sp_io::storage::set(&BOOLEAN_KEY, &next_state.encode());
+			Ok(Ok(()))
+		}
+
+		fn finalize_block() -> <Block as BlockT>::Header {
+			if_std!{
+				println!("Entering finalize_block");
+			}
+			// https://substrate.dev/rustdocs/master/sp_runtime/generic/struct.Header.html
+			let raw_header = sp_io::storage::get(&HEADER_KEY)
+				.expect("We initialized with header, it never got mutated, qed");
+
+			// Clear the raw header out of storage when we are done with it.
+			sp_io::storage::clear(&HEADER_KEY);
+
+			let mut header = <Block as BlockT>::Header::decode(&mut &*raw_header)
+				.expect("we put a valid header in in the first place, qed");
+
+			let raw_state_root = &sp_io::storage::root()[..];
+
+			header.state_root = sp_core::H256::decode(&mut &raw_state_root[..]).unwrap();
+			header
+		}
+
+		// This runtime does not expect any inherents so it does not insert any into blocks it builds.
+		fn inherent_extrinsics(_data: sp_inherents::InherentData) -> Vec<<Block as BlockT>::Extrinsic> {
+			Vec::new()
+		}
+
+		// This runtime does not expect any inherents, so it does not do any inherent checking.
+		fn check_inherents(
+			_block: Block,
+			_data: sp_inherents::InherentData
+		) -> sp_inherents::CheckInherentsResult {
+			sp_inherents::CheckInherentsResult::default()
+		}
+
+		// This runtime does not have a need for a random seed. Nor does it make any effort to
+		// supply a random seed.
+		fn random_seed() -> <Block as BlockT>::Hash {
+			<Block as BlockT>::Hash::from([0u8;32])
+		}
+	}
+
+	impl sp_transaction_pool::runtime_api::TaggedTransactionQueue<Block> for Runtime {
+		fn validate_transaction(
+			_source: TransactionSource,
+			_tx: <Block as BlockT>::Extrinsic) -> TransactionValidity {
+			// Any transaction of the correct type is valid
+			Ok(ValidTransaction{
+				priority: 1u64,
+				requires: Vec::new(),
+				// This hach was necessary to make the node accept any transactions at all
+				// When I was setting provides to an empty vec, submiting a transaction failed
+				// the RPC responded {"code":-32603,"message":"Unknown error occurred","data":"Pool(NoTagsProvided)"}
+				// Adding this provides tag solved that. Solutions moving forward:
+				// 1. Require a nonce with each transaction
+				// 2. Try to relax the TxPool's requirement that every transaction provide some tag
+				provides: vec![vec![0]],
+				longevity: TransactionLongevity::max_value(),
+				propagate: true,
+			})
```

Comment:
@todr The intention of this simple runtime is that transactions can be applied in absolutely any order. There are no dependencies between transactions and no meaningful tags to associate with the transactions. But https://github.com/paritytech/substrate/blob/75113aae02ee3cc265aaaa13cdf5a79c7ff98105/client/transaction-pool/graph/src/pool.rs#L410-L412 enforces that each transaction provide at least one tag. Is there any harm in removing this requirement?

Reply:
Oops... @tomusdrw

Reply:
Sorry, somehow missed the previous mention.

The dependencies are indicated via `requires` tag set, `provides` guarantees de-duplication of the transactions.
Imagine that we allow no `provides` tag. The same transaction can then be included in the pool and in the block as many times as you would like - so each time you receive the transaction over network from your peers it would be imported to the pool and added to the block, so you would end up with infinite number of copies of the same transaction.

I'd suggest adding `provides: vec![hash(tx.encode())]` here. This will de-duplicate transactions floating around the network, but note that it doesn't prevent replaying the same transaction in consecutive blocks.
The runtime must include some kind of replay protection and at some point return `Stale` from `validate_transaction` for transactions that have already been applied in the network.

I'd suggest adding at least `apply_before_block` parameter to each extrinsic, so that we can detect stale transactions. Note that it still does not prevent replaying the same transaction over and over again (in the same block or subsequent blocks) when it's valid.

Reply:
A scheme I had in mind to allow both sequential and parallel transactions some time ago looks like this.
1. Each transaction has `(random_salt, nonce)`.
2. If you want transactions to be included in parallel you set `nonce=0` and choose a random `salt`.
3. If you want transactions to be included sequentially you use the same `salt` and increment the `nonce`.
4. Runtime maintains storage entries for:
```rust
nonce_expirations: DoubleMap<BlockNumber, Salt, ()>,
next_nonce: Map<Salt, (Nonce, BlockNumber)>,
```
Salts are valid for fixed `X` blocks, and every time a salt is used we do something like this:
```
let expire_at = System::current_block_number() + X;
// Update next expected nonce and pruning block.
let (nonce, block) = next_nonce.get(salt).unwrap_or_else(|| (0, 0));
next_nonce.insert(salt, (nonce + 1, expire_at));
// remove previous expiration notice
nonce_expirations.remove((block, salt));
// insert new expiration
nonce_expirations.insert((expire_at, salt), ());
```
And then the runtime can use `next_nonce` to figure out validity of the transaction, and in `on_initialize` we take all `nonce_expirations` at that particular block and prune out all the salts (since we use `DoubleMap` we can get all entries efficiently).